### PR TITLE
Generate Keymap bindings

### LINKED ISSUE #310: Missing gdk_key_map_* functions
I noticed the various functions related to  `gdk_key_map_*` and `gdk_keyval_*`  as described in https://developer.gnome.org/gdk3/stable/gdk3-Keyboard-Handling.html are missing. Is anyone working on this? Is this something I might be able to add in a PR?

### PR DESCRIPTION
**Pull request by bilelmoussaoui**
Fixes #310 

Missing functions:
- [x] get_entries_for_keycode
- [x] get_entries_for_keyval
- [x] lookup_key

The three depend on KeymapKey that causes an issue, see the comment below.

### CODE REVIEW

**Comment by bilelmoussaoui** [comment]
Tried generating Gdk.KeymapKey too but I'm getting this
```
thread 'main' panicked at 'Missing memory management functions for Gdk.KeymapKey', src/codegen/record.rs:75:9
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:39
   1: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:71
   2: std::panicking::default_hook::{{closure}}
             at src/libstd/sys_common/backtrace.rs:59
             at src/libstd/panicking.rs:197
   3: std::panicking::default_hook
             at src/libstd/panicking.rs:211
   4: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:474
   5: std::panicking::continue_panic_fmt
             at src/libstd/panicking.rs:381
   6: std::panicking::begin_panic_fmt
             at src/libstd/panicking.rs:336
   7: libgir::codegen::record::generate
   8: libgir::file_saver::save_to_file
   9: libgir::codegen::records::generate
  10: libgir::codegen::generate
  11: gir::main
  12: std::rt::lang_start::{{closure}}
  13: std::panicking::try::do_call
             at src/libstd/rt.rs:49
             at src/libstd/panicking.rs:293
  14: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:85
  15: std::rt::lang_start_internal
             at src/libstd/panicking.rs:272
             at src/libstd/panic.rs:394
             at src/libstd/rt.rs:48
  16: main
  17: __libc_start_main
  18: _start
make: *** [Makefile:22: src/auto/mod.rs] Error 101
```

**Comment by GuillaumeGomez** [comment]
Which means it needs to be done by hand. :)

**Comment by bilelmoussaoui** [comment]
Oh, will do that :)

**Comment by GuillaumeGomez** [comment]
Look at other similar integer-like structs in the manual bindings. It'll help you to understand how to do it.

**Comment by bilelmoussaoui** [comment]
Ready to be reviewed :)

File: src/auto/keymap_key.rs [hunk 1]
```diff
@@ -0,0 +1,5 @@
+// This file was generated by gir (https://github.com/gtk-rs/gir)
+// from gir-files (https://github.com/gtk-rs/gir-files)
+// DO NOT EDIT
+
+use gdk_sys;
```
Comment (GuillaumeGomez) [inline src/auto/keymap_key.rs]: Considering that this file is mostly empty, no need to generate it.

File: src/keys.rs [hunk 2]
```diff
@@ -7,6 +7,61 @@ use glib::translate::*;
 use glib::GString;
 use libc::c_uint;
 use std::mem::transmute;
+use glib::translate::*;
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+#[repr(C)]
+pub struct KeymapKey {
```
Comment (GuillaumeGomez) [inline src/keys.rs]: Why not putting it into its own file?

File: src/auto/keymap_key.rs (hunk 1 above)
Comment (bilelmoussaoui) [reply src/auto/keymap_key.rs]: Oups, i missed this one. Fixed

File: src/keymap_key.rs [hunk 3]
```diff
@@ -0,0 +1,60 @@
+// Copyright 2013-2015, The Gtk-rs Project Developers.
+// See the COPYRIGHT file at the top-level directory of this distribution.
+// Licensed under the MIT license, see the LICENSE file or <http://opensource.org/licenses/MIT>
+
+use gdk_sys;
+use glib::translate::*;
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+#[repr(C)]
+pub struct KeymapKey {
```
Comment (sdroege) [inline src/keymap_key.rs]: Why not make it a tuple struct around `ffi::GdkKeyMapKey`? Then we don't have to duplicate the struct layout here

**Comment by sdroege** [comment]
Otherwise looks good to me but CI is unhappy

```
error[E0412]: cannot find type `GdkKeymapKey` in this scope
  --> src\keymap_key.rs:53:54
   |
53 | impl FromGlibPtrFull<*mut gdk_sys::GdkKeymapKey> for GdkKeymapKey {
   |                                                      ^^^^^^^^^^^^
help: a struct with a similar name exists
   |
53 | impl FromGlibPtrFull<*mut gdk_sys::GdkKeymapKey> for KeymapKey {
   |                                                      ^^^^^^^^^
help: possible candidate is found in another module, you can import it into scope
   |
5  | use gdk_sys::GdkKeymapKey;
   |
error: Compilation failed, aborting rustdoc
```

**Comment by GuillaumeGomez** [comment]
`get_entries_for_keycode` is badly generated, this is a gir bug. Working on it.

File: src/keymap.rs [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+                &mut keyvals,
+                n_entries.as_mut_ptr(),
+            ));
+            if ret {
+                let n_entries = n_entries.assume_init() as usize;
+                let mut r_keys = Vec::with_capacity(n_entries);
+                let mut r_keyvals = Vec::with_capacity(n_entries);
+                for i in 0..n_entries {
+                    r_keys.push(from_glib_none(keys.add(i)));
+                    r_keyvals.push(ptr::read(keyvals.add(i)));
```
Comment (sdroege) [inline src/keymap.rs]: This seems strange. Why `(Vec<KeymapKey>, Vec<u32>)` and not `Vec<(KeymapKey, u32)>` if they're the same size anyway and apparently belong together? :)

File: src/keymap.rs [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+            let mut keyvals = ptr::null_mut();
+            let mut n_entries = mem::MaybeUninit::uninit();
+            let ret = from_glib(gdk_sys::gdk_keymap_get_entries_for_keycode(
+                self.to_glib_none().0,
+                hardware_keycode,
+                &mut keys,
+                &mut keyvals,
+                n_entries.as_mut_ptr(),
+            ));
+            if ret {
```
Comment (sdroege) [inline src/keymap.rs]: Why not return an empty `Vec` if there were no entries? Less wrapping and that's exactly what a return value of `false` signals: no entries.

File: src/keymap_key.rs (hunk 3 above)
Comment (sdroege) [reply src/keymap_key.rs]: Still a question :)

File: src/modifier_type.rs [hunk 6]
```diff
@@ -0,0 +1,12 @@
+use gdk_sys;
+use glib::translate::ToGlib;
+use ModifierType;
+
+#[doc(hidden)]
+impl ToGlib for &mut ModifierType {
```
Comment (sdroege) [inline src/modifier_type.rs]: Why is this needed?

File: src/auto/keymap.rs [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+
+    match fn {
+        get_type => || gdk_sys::gdk_keymap_get_type(),
+    }
+}
+
+impl Keymap {
+    pub fn add_virtual_modifiers(&self, state: &mut ModifierType) {
+        unsafe {
+            gdk_sys::gdk_keymap_add_virtual_modifiers(self.to_glib_none().0, state.to_glib());
```
Comment (sdroege) [inline src/auto/keymap.rs]: This function sets bits in the `state`, but here you pass a copy of the `state` to the C function and don't read it back afterwards.

File: src/auto/keymap.rs [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+
+    pub fn lookup_key(&self, key: &KeymapKey) -> u32 {
+        unsafe { gdk_sys::gdk_keymap_lookup_key(self.to_glib_none().0, key.to_glib_none().0) }
+    }
+
+    pub fn map_virtual_modifiers(&self, state: &mut ModifierType) -> bool {
+        unsafe {
+            from_glib(gdk_sys::gdk_keymap_map_virtual_modifiers(
+                self.to_glib_none().0,
+                state.to_glib(),
```
Comment (sdroege) [inline src/auto/keymap.rs]: Also wrong like `add_virtual_modifiers()` with regards to the `state`

File: src/keymap.rs (hunk 4 above)
Comment (GuillaumeGomez) [reply src/keymap.rs]: Just thought about it afterwards. :)

File: src/modifier_type.rs [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+    type GlibType = *mut gdk_sys::GdkModifierType;
+
+    fn to_glib(&self) -> *mut gdk_sys::GdkModifierType {
+        &mut self.bits() as *mut _
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/modifier_type.rs]: How would you do it?

File: src/auto/keymap.rs (hunk 7 above)
Comment (GuillaumeGomez) [reply src/auto/keymap.rs]: Based on this code: https://code.woboq.org/gtk/gtk/gdk/x11/gdkkeys-x11.c.html#gdk_x11_keymap_add_virtual_modifiers, it updates the `state` so I'm not sure what you're saying. It seems good to me?

File: src/modifier_type.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (sdroege) [reply src/modifier_type.rs]: Remove this :)

File: src/auto/keymap.rs (hunk 7 above)
Comment (sdroege) [reply src/auto/keymap.rs]:
```rust
    pub fn add_virtual_modifiers(&self, state: &mut ModifierType) {
        unsafe {
            let mut s = state.to_glib();
            gdk_sys::gdk_keymap_add_virtual_modifiers(self.to_glib_none().0, &mut s);
            *state = from_glib(state);
       }
    }
```

Comment (GuillaumeGomez) [reply src/auto/keymap.rs]: Oh I see! Thanks!

**Comment by GuillaumeGomez** [comment]
Updated.

**Comment by sdroege** [comment]
:+1:

**Comment by GuillaumeGomez** [comment]
Finally! \o/

**Comment by bilelmoussaoui** [comment]
Thanks @GuillaumeGomez for taking care of this
